File: test-suite/src/test/java/io/micronaut/security/oauth2/docs/endpoint/OAuthCsrfFilter.java
package io.micronaut.security.oauth2.docs.endpoint;

//tag::imports[]
import io.micronaut.context.annotation.Requires;
import io.micronaut.core.async.publisher.Publishers;
import io.micronaut.http.HttpRequest;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.HttpStatus;
import io.micronaut.http.MutableHttpResponse;
import io.micronaut.http.annotation.Filter;
import io.micronaut.http.cookie.Cookie;
import io.micronaut.http.filter.HttpServerFilter;
import io.micronaut.http.filter.ServerFilterChain;
import org.reactivestreams.Publisher;

//end::imports[]
@Requires(property = "oauth.csrf")
//tag::class[]
@Filter(value = {"/oauth/login", "/oauth/login/*"})
public class OAuthCsrfFilter implements HttpServerFilter {

    @Override
    public Publisher<MutableHttpResponse<?>> doFilter(HttpRequest<?> request, ServerFilterChain chain) {
        String requestParameter = request.getParameters().get("_csrf");
        String cookieValue = request.getCookies().findCookie("_csrf").map(Cookie::getValue).orElse(null);

        if (cookieValue == null || !cookieValue.equals(requestParameter)) {
            return Publishers.just(HttpResponse.status(HttpStatus.FORBIDDEN));
        }

        return chain.proceed(request);
    }
}
//end::class[]

File: test-suite-python/src/test/python/micronaut/security/oauth2/docs/endpoint/OAuthCsrfFilter.py
# tag::imports[]
from micronaut.context.annotation import Requires
from micronaut.core.async_.publisher import Publishers
from micronaut.http import HttpRequest, HttpResponse, HttpStatus, MutableHttpResponse
from micronaut.http.annotation import Filter
from micronaut.http.filter import HttpServerFilter, ServerFilterChain
from org.reactivestreams import Publisher
# end::imports[]


@Requires(property="oauth.csrf")
# tag::class[]
@Filter(value=["/oauth/login", "/oauth/login/*"])
class OAuthCsrfFilter(HttpServerFilter):

    def doFilter(self, request: HttpRequest, chain: ServerFilterChain) -> Publisher[MutableHttpResponse]:
        requestParameter = request.getParameters().get("_csrf")
        cookieValue = request.getCookies().findCookie("_csrf").map(lambda cookie: cookie.getValue()).orElse(None)

        if cookieValue is None or cookieValue != requestParameter:
            return Publishers.just(HttpResponse.status(HttpStatus.FORBIDDEN))

        return chain.proceed(request)
# end::class[]
